// Google Protocol Buffers used for RPCs between servers and clients.
syntax = "proto3";

message AppendEntries {
    /* SHARED FIELD */
    int32 term = 1;

    /* REQUEST FIELD */
    int32 leader_no = 2;

    /* RESPONSE FIELDS */
    int32 follower_no = 3;
    bool success = 4;
    int32 index_last_commit = 5;
}

message RequestVote {
    /* SHARED FIELD */
    int32 term = 1;

    /* REQUEST FIELD */
    int32 candidate_no = 2;

    /* RESPONSE FIELDS */
    int32 voter_no = 3;
    bool vote_granted = 4;
}

message ClientRequest {
    string command = 1;
    string client_HostAndPort = 2; // so server knows who to repond to. 
                                  // will be removed after socket architecture change

    /* RESPONSE FIELDS */
    bool success = 4;
    int32 leader_id = 5;
    string output = 6;
}

message RAFTmessage {
    int32 term = 1;
    // Message instances
    AppendEntries AppendEntries_message = 2;
    RequestVote RequestVote_message = 3;
    ClientRequest ClientRequest_message = 4;
}

// For servers writing their persistent state variables to disk.
message ServerPersistentState {
    int32 current_term = 1;
    int32 term_voted = 2;
    int32 voted_for = 3;
}
